WIP feat(routes): add support for route middleware

hypeJunction opened the pull request:
Plugins can now define middleware to be executed before route is rendered.
Makes it simpler to allow plugins to define public routes available
when walled garden is enabled.

- [ ] Needs #11687

hypeJunction committed:
feat(routes): add support for route middleware

Plugins can now define middleware to be executed before route is rendered.
Makes it simpler to allow plugins to define public routes available
when walled garden is enabled.

hypeJunction commented on engine/classes/Elgg/WalledGatdenException.php:
```diff
@@ -0,0 +1,10 @@
+/**
+ * Thrown when walled garden gatekeeper prevents access
+ */
+class WalledGatdenException extends GatekeeperException {
```
Garden that is

jeabakker commented on docs/guides/routing.rst:
```diff
@@ -162,6 +162,59 @@ By default, Elgg will set the following requirements for named URL segments:
+Core provides 3 middleware handlers: ``Gatekeeper``, ``AdminGatekeeper``, ``WalledGarden``. Plugins can implement their
+own middleware by extending ``\Elgg\Router\Middleware`` class.
+
+.. php::code
```
almost should be `.. code-block:: php`

jeabakker commented on engine/classes/Elgg/Router.php:
```diff
@@ -260,9 +289,13 @@ public function registerPageHandler($identifier, $function) {
 		$resource = elgg_extract('resource', $params);
 		$handler = elgg_extract('handler', $params);
+		$middleware = elgg_extract('middleware', $params, []);
+		$public = elgg_extract('public', $params, false);
```
why default to non-public routes, doesn't that cause lots of problems

hypeJunction replied:
That's the current approach. When you enable WG everything is non-public, unless allowed by hook

jeabakker replied:
ah so this only gets enforced by WG (and added middleware)?

hypeJunction replied:
I suppose "public" is ambiguous. Should rename into something else. Ideas?

jeabakker commented:
Maybe "guarded" or "gated" I don't have anything better

Op 31 jan. 2018 16:36 schreef "Ismayil Khayredinov" <
notifications@github.com>:

> *@hypeJunction* commented on this pull request.
> ------------------------------
>
> In engine/classes/Elgg/Router.php
> <https://github.com/Elgg/Elgg/pull/11682#discussion_r165090237>:
>
> >  		$path = elgg_extract('path', $params);
>  		$resource = elgg_extract('resource', $params);
>  		$handler = elgg_extract('handler', $params);
> +		$middleware = elgg_extract('middleware', $params, []);
> +		$public = elgg_extract('public', $params, false);
>
> I suppose "public" is ambiguous. Should rename into something else. Ideas?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Elgg/Elgg/pull/11682#discussion_r165090237>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA11JlwavnkPCG3oqWJeLd60RGyXQzf2ks5tQIhrgaJpZM4Rz3pR>
> .
>

hypeJunction committed:
feat(di): make session a public service

Elgg session is now a public service available via elgg()->session

hypeJunction committed:
feat(di): make gatekeeper a public service

Gatekeeper is now accessible via elgg()->gatekeeper.

hypeJunction committed:
feat(di): add translator to public services

Translator is now avaialable via elgg()->translator.
elgg_echo() is now aliases as elgg()->echo().

hypeJunction committed:
feat(di): add hooks and events to public services

elgg()->hooks and elgg()->events can now be used to access
public services

hypeJunction committed:
feat(di): add system messages to public services

elgg()->system_messages can now be used to access system messages
service

hypeJunction committed:
feat(di): add config to public services

hypeJunction commented:
Fixes #11177

hypeJunction commented:
We will need to eventually kill context and input altogether. Both are stupid and volatile, but this will do for now. Request will need to be thrown out of services as it is not a service, it needs to be property of the router after being routed. Route registration needs to move to it's own service.

hypeJunction committed:
chore(routes): add dedicated route registration service

Decouples router from route registration

hypeJunction committed:
feat(router): add support for controllers

Routes can now be handled with controllers, instead of resource views.

hypeJunction committed:
feat(route): add support for handling routes using files

Files can now be used as route controllers

hypeJunction commented:
Once this is in, there is another PR coming with rewiring actions system to use routes and middleware. Almost done there, just need to update the tests. And we will finally have a normal uniform OOP routing system

hypeJunction commented:
Replaced by #11690
